clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java

```diff
@@ -2302,6 +2317,49 @@ public void testRebalanceMetricsOnFailedRebalance() {
         assertEquals(-1d, getMetricValue(metrics, rebalanceMetricsManager.lastRebalanceSecondsAgo));
     }
 
+    @Test
+    public void testLeaveGroupWhenStateIsFatal() {
+        MembershipManagerImpl membershipManager = createMemberInStableState(null);
+        when(subscriptionState.hasAutoAssignedPartitions()).thenReturn(true);
+        membershipManager.transitionToFatal();
+        assertEquals(MemberState.FATAL, membershipManager.state());
+
+        membershipManager.leaveGroup();
+        verify(subscriptionState).unsubscribe();
+    }
+
+    @Test
+    public void testLeaveGroupWhenStateIsUnsubscribe() {
+        MembershipManagerImpl membershipManager = new MembershipManagerImpl(
+                GROUP_ID, Optional.empty(), REBALANCE_TIMEOUT, Optional.empty(),
+                subscriptionState, commitRequestManager, metadata, LOG_CONTEXT, Optional.empty(),
+                backgroundEventHandler, time, rebalanceMetricsManager);
+        assertEquals(MemberState.UNSUBSCRIBED, membershipManager.state());
+
+        membershipManager.leaveGroup();
+        verify(subscriptionState).unsubscribe();
+    }
+
+    @Test
+    public void testLeaveGroupWhenStateIsFenced() {
+        MembershipManagerImpl membershipManager = createMemberInStableState();
+        ConsumerRebalanceListenerInvoker invoker = consumerRebalanceListenerInvoker();
+        mockFencedMemberStuckOnUserCallback(membershipManager, invoker);
+        assertEquals(MemberState.FENCED, membershipManager.state());
+
+        membershipManager.leaveGroup();
+        verify(subscriptionState).unsubscribe();
+    }
+
+    @Test
+    public void testLeaveGroupWhenStateIsStale() {
```

Comment:
sorry a bit late, but I noticed now that this test is a duplicate, we already have a `testLeaveGroupWhenMemberIsStale`, that you updated to verify the unsubscribe, so I would say we just remove this one.

Reply:
Agree. Removed them. Thanks for review and finding this. 👍

Reply:
one last: the `testLeaveGroupWhenStateIsUnsubscribe` could be removed too, we already have `testLeaveGroupWhenMemberAlreadyLeft` for that. (Let's keep the `testLeaveGroupWhenStateIsFatal`, we didn't have coverage for that before this PR, thanks!)

Reply:
> Let's keep the testLeaveGroupWhenStateIsFatal, we didn't have coverage for that before this PR

It seems to me `testLeaveGroupWhenStateIsFatal` is used to verify "leaveGroup" after it is in fatal state. There are many existent cases which can offer the scenario. For example: `testFatalFailureWhenMemberAlreadyLeft`. Hence, we can add `leaveGroup` and `verify(subscriptionState)` to those test cases after the state is in fatal

Reply:
Hi @lianetm and @chia7712, thanks for your patience and review. I remove both `testLeaveGroupWhenStateIsUnsubscribe` and `testLeaveGroupWhenStateIsFatal`. I add `leaveGroup` test to `testTransitionToFatal`.